# **Data cleaning and preparation for financial analysis**

##### This notebook was created to show how I prepared a dataset to analyze my financial spending data, using credit card, for the years 2018, 2019 and 2020. To analyze the 3 years of my personal sppending, I had to unify the 3 differente files in one. The first problem: I updated almost twice a year the excel file to optimize record information and analysis. So, I had to read each year file and see the differences in each month, correct the differences and create a standard table. Then I created a unique file to join the 3 years.


##### **The operations applied on the process are as shown below:**

1. Creating a Data Frame
2. Performing operations on Rows and Columns
3. Data Selection, addition, deletion
4. Exporting a CSV file
5. Exporting Excel file
6. Unifying the 3 excel files (2018, 2019 e 2020)
7. Using Excel to correct data

## 1.Creating a Data Frame

##### The first pandas operation was read an excel file of 2018. This operation returns a dict which each item corresponds to a sheet, this one is the data frame of each month.

In [1]:
# Import pandas library
import pandas as pd

# Creating a dict of data frames
dictdf = pd.read_excel('1.2018.Tables.xlsx', None)
# dictdf

# 2.Performing operations on Rows and Columns

##### The name of sheets in the file is the full name of the respective month (in portuguese), so I created a list with month names for further operations.

In [2]:
# Creating the list of the names
months = []

# Inserting the sheet names (months) in the list by using a loop.
for m in dictdf.keys():
    months.append(m)
    print(m, end=", ")

# months

Janeiro, Fevereiro, Março, Abril, Maio, Junho, Julho, Agosto, Setembro, Outubro, Novembro, Dezembro, 

In [3]:
# Using shape method to see the differences between sheets.
for m in months:
    print(dictdf[m].shape)

(30, 10)
(56, 10)
(35, 10)
(40, 10)
(41, 10)
(55, 10)
(53, 10)
(64, 10)
(54, 10)
(68, 10)
(131, 9)
(92, 9)


##### *Note that the 2 last months shows a higher number of rows and 1 less column.*

##### This is because in the first file model I was recording the spending by credit in a table separated from spending by debit/cash. After I update the file, I started recording both (credit and debit) in a single table, which I need to arrange to be able to joint the sheets.

In [4]:
# Printing column name of each sheet to see what need to be change
for m in months:
    print(dictdf[m].columns)    

Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
Index(['Local', 

## 3.Data Selection, addition, deletion.

##### The last 2 months have 'Modo' column, which defines the spending type (credit or debit) and don't have the columns 'Mês' e 'Ano' ('Month' and 'Year', respectively).

##### Since my goal is to analyse spending by credit, the first change I made was to drop rows with debit type. And I took the opportunity to add 'Mês' and 'Ano' columns and drop the 'Modo' column, once I don't need it anymore.

In [5]:
#Using enumarate to get the index and the name of the month.
for i, m in enumerate(months):
    #Since I want just the last two months, I set the index for greater than 9.
    if i > 9:        
        # Droping the rows and inserting columns.
        dictdf[m].drop(dictdf[m][dictdf[m]['Modo'] == 'Débito'].index, inplace = True)
        dictdf[m].insert(3, 'Mês', i+1)
        dictdf[m].insert(4, 'Ano', 2018)
        dictdf[m].drop('Modo', axis =1, inplace = True)
    

In [6]:
# Verifying column name of each sheet to see if it's the same and its shape.
for m in months:
    print(dictdf[m].columns)   
    print(dictdf[m].shape)

Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(30, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(56, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(35, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(40, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(41, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(55, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
       'Cartão', 'Valor'],
      dtype='object')
(53, 10)
Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Tipo',
  

##### Now all sheet has the same columns names and quantity, but the month number and year aren't right (only on the 2 last months). So I fixed it.

In [7]:
for i, m in enumerate(months):
    if i < 10:
        for c in range(len(dictdf[m])):
            dictdf[m].loc[c, 'Mês'] = (i + 1)
            dictdf[m].loc[c, 'Ano'] = 2018
            

##### Now I can concatenate all sheets in just one dataframe.

In [8]:
# Using concat to concatenate all sheets in one dataframe
df = pd.concat(dictdf[sheet] for sheet in dictdf.keys())

In [9]:
# Droping rows which 'Local' column is a NaN.
df.dropna(subset=['Local'], inplace=True)

In [10]:
# Droping rows which value is less than 0.
df.drop(df[df['Valor'] < 0].index, inplace=True)
df.shape

(598, 10)

# 5. Exporting Excel file.

##### The 2018 dataframe end up with 598 rows and 10 columns. Now I can prepare the data from next years and than concatenate all of them in a single sheet do be able do analyse in visualization software.

##### I didn't record the process of 2019 and 2020. So in the next topic I show how I cleand up to remove information that was not needed to create the final file.

In [12]:
df.to_excel('1.2018.Clean.xlsx')

# 6. Unifying the 3 excel files (2018, 2019 e 2020).

In [13]:
import pandas as pd
# Reading all tree files in differents dataframes
df2018 = pd.read_excel('1.2018.Clean.xlsx')
df2019 = pd.read_excel('2.2019.Clean.xlsx')
df2020 = pd.read_excel('3.2020.Clean.xlsx')

In [14]:
# Creating a list with the 3 tables
dfs = [df2018[:], df2019[:], df2020[:]]

In [15]:
# Loop to see the column names
for i in dfs:
    print(i.columns)

Index(['ID', 'Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo',
       'Categoria', 'Tipo', 'Divisão Salarial', 'Cartão', 'Valor'],
      dtype='object')
Index(['ID', 'Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo',
       'Categoria', 'Tipo', 'Divisão Salarial', 'Cartão', 'Valor'],
      dtype='object')
Index(['ID', 'Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo',
       'Categoria', 'Tipo', 'Divisão Salarial', 'Cartão', 'Valor'],
      dtype='object')


In [16]:
# Change de name that was created when I exported the files in ID columns
for i in range(3):
    dfs[i].rename(columns = {'Unnamed: 0':'ID'}, inplace = True)

In [17]:
# Inserting columns 'Tipo' and 'Divisão Salarial' in 2018 table
dfs[0].insert(9, 'Tipo', '')
dfs[0].insert(10, 'Divisão Salarial', '')

In [18]:
# Here I tried do reindex the ID, but I couldn't. So I exported the dataframe to do it. 
fdf = pd.concat(dfs[d][:] for d in range(3))
fdf.to_excel('4.FinalDataFrame.xlsx')

In [19]:
# Just reading the same dataframe
fdf = pd.read_excel('4.FinalDataFrame.xlsx')
print(fdf.columns)
fdf.shape

Index(['Unnamed: 0', 'Unnamed: 0.1', 'ID', 'Local', 'Pessoa', 'Data', 'Mês',
       'Ano', 'Ítem', 'Motivo', 'Categoria', 'Tipo', 'Divisão Salarial',
       'Cartão', 'Valor'],
      dtype='object')


(1352, 15)

In [21]:
# Dropping the columns that was created by exporting
fdf.drop(['Unnamed: 0', 'Unnamed: 0.1'], axis=1, inplace = True)
print(fdf.columns)
fdf.shape

Index(['ID', 'Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo',
       'Categoria', 'Tipo', 'Divisão Salarial', 'Cartão', 'Valor'],
      dtype='object')


(1352, 13)

In [23]:
# Counting the number of rows
fdf.count()

ID                  1352
Local               1352
Pessoa              1352
Data                1352
Mês                 1352
Ano                 1352
Ítem                1352
Motivo              1352
Categoria           1352
Tipo                1352
Divisão Salarial    1352
Cartão              1330
Valor               1352
dtype: int64

In [25]:
# Dropping rows with no valid values
fdf.drop(fdf[fdf['Pessoa'] != 'Eu'].index, inplace=True)
fdf.drop(fdf[fdf['Valor'] <= 0].index, inplace=True)
fdf.dropna(subset=['Valor'], inplace=True)

In [26]:
# Checking the numbers
fdf.count()

ID                  1352
Local               1352
Pessoa              1352
Data                1352
Mês                 1352
Ano                 1352
Ítem                1352
Motivo              1352
Categoria           1352
Tipo                 929
Divisão Salarial     742
Cartão              1330
Valor               1352
dtype: int64

In [28]:
# Counting the numbers os rows with the same category
dupcat = fdf.groupby('Categoria')['Categoria'].count().sort_values(ascending  = False)
dupcat = dupcat.to_dict()

In [29]:
for i, k in dupcat.items():
    print(f'{i:<20} {k}')

Combustível          190
Restaurante          151
Alimentação          91
Lazer                76
Presente             73
Eletrônico           70
Roupas               69
Saúde                61
Profissional         48
Viagem               38
Dança                35
UBER                 33
Celular              33
Higiene              30
Deslocamento         30
Estudo               30
Carro                30
Estacionamento       26
Bicicleta            21
Contas               20
Transporte           19
Lanches              18
Streaming            14
Capoeira             14
Lanche               10
Combustítvel         9
Cash Back            9
Passagem             9
Educação             7
Combústivel          6
Quarto               6
 Presente            5
Necessidade Básica   5
Internet             5
Intercâmbio          4
Cachorro             4
Dog                  4
TCC                  4
Jogo                 4
Necessidades Básicas 3
Estética             3
Jogos                3
Pet    

In [30]:
fdf.to_excel('4.FinalDataFrame.xlsx')

##### Here I decided to use excel to correct the columns 'Categoria', 'Tipo' e 'Divisão Salarial' because it's a personal sheet, so I preferred to look more closely, sometimes at each cell, and change it as I thought necessary. 

# 7. Uying Excel to correct data

##### In the column 'Categoria' I looked up for same categories that had different names or the name was just typed wrong. In addition, I completed the column 'Tipe' based on the date and the item description. The column 'Salary Division' a correct based on the category (column 'Categoria'). I used commands like VLOOKUP and COUNTIFS in this process.

In [31]:
# Reading excel file after changes.
fdf = pd.read_excel('4.FinalDataFrameExcel.xlsx')
print(fdf.columns)
fdf.shape

Index(['Local', 'Pessoa', 'Data', 'Mês', 'Ano', 'Ítem', 'Motivo', 'Categoria',
       'Tipo', 'Divisão Salarial', 'Cartão', 'Valor'],
      dtype='object')


(1352, 12)

In [33]:
#Verifying number os rows in each column.
fdf.count()

Local               1352
Pessoa              1352
Data                1352
Mês                 1352
Ano                 1352
Ítem                1352
Motivo              1352
Categoria           1352
Tipo                1352
Divisão Salarial    1352
Cartão              1330
Valor               1352
dtype: int64

##### To get the answers I need in my analysis, I’m only going to use the columns 'Data', 'Mês', 'Ano', 'Categoria', 'Tipo', 'Divisão Salarial', 'Valor'. So I can drop the rest.

In [35]:
#Droppin columns that I don't need anymore.

fdf.drop('Local', axis=1, inplace = True)
fdf.drop('Pessoa', axis=1, inplace = True)
fdf.drop('Cartão', axis=1, inplace = True)
fdf.drop('Ítem', axis=1, inplace = True)
fdf.drop('Motivo', axis=1, inplace = True)
print(fdf.columns)
fdf.shape

Index(['Data', 'Mês', 'Ano', 'Categoria', 'Tipo', 'Divisão Salarial', 'Valor'], dtype='object')


(1352, 7)

##### Now my table is clean and prepared. I'm able to do my analysis.

In [37]:
fdf.to_excel('4.FinalTable.xlsx')